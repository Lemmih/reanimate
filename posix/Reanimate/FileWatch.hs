module Reanimate.FileWatch
  ( watchFile ) where

import Control.Concurrent
import           System.Directory

watchFile :: FilePath -> IO () -> IO ()
watchFile path handler = do
    t0 <- getModificationTime path
    forkIO $ loop t0
    return ()
  where
    loop lastModTime = do
      t1 <- getModificationTime path
      if t1 /= lastModTime
        then handler >> loop t1
        else threadDelay (10^3 * delay) >> loop lastModTime
    delay = 1000 -- pool delay in ms.
