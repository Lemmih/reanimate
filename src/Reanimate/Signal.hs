module Reanimate.Signal
  ( Signal
  , signalFlat
  , signalLinear
  , signalFromTo
  , signalReverse
  , signalCurve
  , signalBell
  , signalOscillate
  , signalFromList
  ) where

type Signal = Double -> Double

signalFromList :: [(Double, Signal)] -> Signal
signalFromList fns t = worker 0 fns
  where
    worker _ [] = 0
    worker now [(len, fn)] = fn (min 1 ((t-now) / min (1-now) len))
    worker now ((len, fn):rest)
      | now+len < t = worker (now+len) rest
      | otherwise = fn ((t-now) / len)

signalFlat :: Double -> Signal
signalFlat x = const x

signalLinear :: Signal
signalLinear = id

signalFromTo :: Double -> Double -> Signal -> Signal
signalFromTo from to c t = from + (to-from)*(c t)

signalReverse :: Signal -> Signal
signalReverse fn t = fn (1-t)

signalCurve :: Double -> Signal
signalCurve steepness s =
  if s < 0.5
    then 0.5 * (2*s)**steepness
    else 1-0.5 * (2 - 2*s)**steepness

signalOscillate :: Signal -> Signal
signalOscillate fn t =
  if t < 1/2
    then fn (t*2)
    else fn (2-t*2)

signalBell :: Double -> Signal
signalBell = signalOscillate . signalCurve
