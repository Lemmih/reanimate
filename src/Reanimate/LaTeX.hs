{-# LANGUAGE OverloadedStrings   #-}
{-# LANGUAGE ScopedTypeVariables #-}
module Reanimate.LaTeX (latex,xelatex,latexAlign) where

import qualified Data.ByteString   as B
import           Data.Monoid ((<>))
import           Data.Text         (Text)
import qualified Data.Text         as T
import qualified Data.Text.IO      as T
import           Graphics.SvgTree  (Tree (..), parseSvgFile)
import           Reanimate.Cache
import           Reanimate.Misc
import           Reanimate.Svg
import           Reanimate.Parameters
import           System.FilePath   (replaceExtension, takeFileName, (</>))
import           System.IO.Unsafe  (unsafePerformIO)

-- | Invoke latex and import the result as an SVG object. SVG objects are
--   cached to improve performance.
--
--   Example:
--
--   > latex "$e^{i\\pi}+1=0$"
--
--   <<docs/gifs/doc_latex.gif>>
latex :: T.Text -> Tree
latex tex | pNoExternals = mkText tex
latex tex = (unsafePerformIO . (cacheMem . cacheDiskSvg) (latexToSVG "dvi" exec args)) script
  where
    exec = "latex"
    args = []
    script = mkTexScript exec args [] tex

-- | Invoke xelatex and import the result as an SVG object. SVG objects are
--   cached to improve performance. Xelatex has support for non-western scripts.
--
--   Example:
--
--   > xelatex "中文"
--
--   <<docs/gifs/doc_xelatex.gif>>
xelatex :: Text -> Tree
xelatex tex | pNoExternals = mkText tex
xelatex tex = (unsafePerformIO . (cacheMem . cacheDiskSvg) (latexToSVG "xdv" exec args)) script
  where
    exec = "xelatex"
    args = ["-no-pdf"]
    headers = ["\\usepackage[UTF8]{ctex}"]
    script = mkTexScript exec args headers tex

-- | Invoke latex and import the result as an SVG object. SVG objects are
--   cached to improve performance. This wraps the TeX code in an 'align*'
--   context.
--
--   Example:
--
--   > latexAlign "R = \\frac{{\\Delta x}}{{kA}}"
--
--   <<docs/gifs/doc_latexAlign.gif>>
latexAlign :: Text -> Tree
latexAlign tex = latex $ T.unlines ["\\begin{align*}", tex, "\\end{align*}"]

postprocess :: Tree -> Tree
postprocess = lowerTransformations . scaleXY 1 (-1) . scale 0.1 . pathify

-- executable, arguments, header, tex
latexToSVG :: String -> String -> [String] -> Text -> IO Tree
latexToSVG dviExt latexExec latexArgs tex = do
  latexBin <- requireExecutable latexExec
  dvisvgm <- requireExecutable "dvisvgm"
  withTempDir $ \tmp_dir -> withTempFile "tex" $ \tex_file -> withTempFile "svg" $ \svg_file -> do
    let dvi_file = tmp_dir </> replaceExtension (takeFileName tex_file) dviExt
    T.writeFile tex_file tex
    runCmd latexBin (latexArgs ++ ["-interaction=nonstopmode", "-halt-on-error", "-output-directory="++tmp_dir, tex_file])
    runCmd dvisvgm [ dvi_file, "--precision=5"
                   , "--exact"    -- better bboxes.
                   -- , "--bbox=1,1" -- increase bbox size.
                   , "--no-fonts" -- use glyphs instead of fonts.
                   ,"--verbosity=0", "-o",svg_file]
    svg_data <- B.readFile svg_file
    case parseSvgFile svg_file svg_data of
      Nothing  -> error "Malformed svg"
      Just svg -> return $ postprocess $ unbox $ replaceUses svg

mkTexScript :: String -> [String] -> [Text] -> Text -> Text
mkTexScript latexExec latexArgs texHeaders tex = T.unlines $
  [ "% " <> T.pack (unwords (latexExec:latexArgs))
  , "\\documentclass[preview]{standalone}"
  , "\\usepackage{amsmath}"
  , "\\usepackage{gensymb}"
  ] ++ texHeaders ++
  [ "\\usepackage[english]{babel}"
  , "\\linespread{1}"
  , "\\begin{document}"
  , tex
  , "\\end{document}" ]

{- Packages used by manim.

\\\usepackage{amsmath}\n\
\\\usepackage{amssymb}\n\
\\\usepackage{dsfont}\n\
\\\usepackage{setspace}\n\
\\\usepackage{relsize}\n\
\\\usepackage{textcomp}\n\
\\\usepackage{mathrsfs}\n\
\\\usepackage{calligra}\n\
\\\usepackage{wasysym}\n\
\\\usepackage{ragged2e}\n\
\\\usepackage{physics}\n\
\\\usepackage{xcolor}\n\
\\\usepackage{textcomp}\n\
\\\usepackage{xfrac}\n\
\\\usepackage{microtype}\n\
-}
