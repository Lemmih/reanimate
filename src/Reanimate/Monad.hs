module Reanimate.Monad where

import           Control.Arrow ()
import qualified Control.Category      as C
import           Control.Monad.State
import           Data.Fixed
import           Data.Fixed            (mod')
import qualified Data.Map              as M
import           Data.Monoid           ((<>))
import           Data.Text             (Text, pack)
import           Graphics.Svg          (Document (..), Number (..), Tree,
                                        xmlOfDocument, Tree(..), Text(..), TextSpan(..), TextSpanContent(..))
import           Reanimate.LaTeX
import           Reanimate.Svg
import           Text.XML.Light        (elContent)
import           Text.XML.Light.Output

import           Reanimate.Combinators (approxFnData, morphPath)

type Duration = Double
type Time = Double

data Frame a = Frame {unFrame :: Duration -> Time -> State ([Tree] -> [Tree]) a}

instance Functor Frame where
  fmap fn f = Frame $ \d t -> fmap fn (unFrame f d t)

instance Applicative Frame where
  pure a = Frame $ \_ _ -> pure a
  fn <*> fa = Frame $ \d t -> do
    fn <- unFrame fn d t
    a <- unFrame fa d t
    pure (fn a)

instance Monad Frame where
  return a = Frame $ \_ _ -> pure a
  f >>= g = Frame $ \d t -> do
    a <- unFrame f d t
    unFrame (g a) d t

-- End behavior:
--   Freeze at last frame
--   Loop
--   Disappear
data Animation = Animation Duration (Frame ())

mkAnimation :: Duration -> Frame () -> Animation
mkAnimation = Animation

duration :: Animation -> Duration
duration (Animation d _) = d

emit :: Tree -> Frame ()
emit svg = Frame $ \_ _ -> modify (.(svg:))

before :: Animation -> Animation -> Animation
before (Animation d1 (Frame f1)) (Animation d2 (Frame f2)) =
  Animation (d1+d2) (Frame $ \_ t -> if t < d1 then f1 d1 t else f2 d2 (t-d1))

-- Play two animation concurrently. Shortest animation freezes on last frame.
sim :: Animation -> Animation -> Animation
sim (Animation d1 (Frame f1)) (Animation d2 (Frame f2)) =
  Animation (max d1 d2) $ Frame $ \d t -> do
    f1 d1 (min d1 t)
    f2 d2 (min d2 t)

-- Play two animation concurrently. Shortest animation loops.
simLoop :: Animation -> Animation -> Animation
simLoop (Animation d1 (Frame f1)) (Animation d2 (Frame f2)) =
  Animation (max d1 d2) $ Frame $ \d t -> do
    f1 d1 (t `mod'` d1)
    f2 d2 (t `mod'` d2)

-- Play two animation concurrently. Animations disappear after playing once.
simDrop :: Animation -> Animation -> Animation
simDrop (Animation d1 (Frame f1)) (Animation d2 (Frame f2)) =
  Animation (max d1 d2) $ Frame $ \d t -> do
    when (t < d1) (f1 d1 t)
    when (t < d2) (f2 d2 t)

pause :: Double -> Animation
pause d = Animation d (pure ())

andThen :: Animation -> Animation -> Animation
andThen a b = a `sim` (pause (duration a) `before` b)

signal :: Double -> Double -> Frame Double
signal from to = Frame $ \d t -> pure $
  from + (to-from)*(t/d)

signalSCurve :: Double -> Double -> Double -> Frame Double
signalSCurve steepness from to = do
  s <- signal 0 1
  let s' = if s < 0.5
              then 0.5 * (2*s)**steepness
              else 1-0.5 * (2 - 2*s)**steepness
  pure $ from + (to-from)*s'

frameAt :: Double -> Animation -> Tree
frameAt t (Animation d (Frame f)) = mkGroup $ execState (f d (min d t)) id []

renderTree :: Tree -> String
renderTree = renderSizedTree Nothing Nothing

renderSizedTree :: Maybe Number -> Maybe Number -> Tree -> String
renderSizedTree w h t = ppElement $ xmlOfDocument doc
  where
    width = 320
    height = width / (16/9)
    doc = Document
      { _viewBox = Just (-width/2, -height/2, width, height)
      , _width = w
      , _height = h
      , _elements = [t]
      , _definitions = M.empty
      , _description = ""
      , _styleRules = []
      , _documentLocation = ""
      }

mapA :: (Tree -> Tree) -> Animation -> Animation
mapA fn (Animation d f) = Animation d (mapF fn f)

mapF :: (Tree -> Tree) -> Frame a -> Frame a
mapF fn frame = Frame $ \d t -> do
  case runState (unFrame frame d t) id of
    (a, children) -> modify (. (fn (mkGroup (children [])):)) >> pure a

pauseAtEnd :: Double -> Animation -> Animation
pauseAtEnd p a = a `andThen` pause p

adjustSpeed :: Double -> Animation -> Animation
adjustSpeed factor (Animation d fn) =
  Animation (d/factor) $ Frame $ \_dur t -> unFrame fn d (t*factor)

reverseAnimation :: Animation -> Animation
reverseAnimation (Animation d fn) = Animation d $ Frame $ \_dur t ->
  unFrame fn d (d-t)

autoReverse :: Animation -> Animation
autoReverse a = a `before` reverseAnimation a

oscillate :: Frame a -> Frame a
oscillate f = Frame $ \d t -> do
  if t < d/2
    then unFrame f d (t*2)
    else unFrame f d (d*2-t*2)

repeatAnimation :: Double -> Animation -> Animation
repeatAnimation n (Animation d f) = Animation (d*n) $ Frame $ \_ t ->
  unFrame f d (t `mod'` d)
