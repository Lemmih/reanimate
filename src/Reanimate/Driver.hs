module Reanimate.Driver ( reanimate ) where

import           Control.Concurrent (MVar, forkIO, killThread, modifyMVar_,
                                     newEmptyMVar, putMVar)
import           Control.Monad.Fix  (fix)
import qualified Data.Text          as T
import           Network.WebSockets
import           System.Directory   (findFile, listDirectory)
import           System.Environment (getArgs, getProgName)
import           System.INotify     (EventVariety (..), addWatch, withINotify)
import           System.IO          (BufferMode (..), hPutStrLn, hSetBuffering,
                                     stderr, stdin)

import           Reanimate.Misc     (runCmdLazy, runCmd, runCmd_, withTempDir, withTempFile)
import           Reanimate.Monad    (Animation)
import           Reanimate.Render   (renderSvgs)

import Paths_reanimate

opts = defaultConnectionOptions
  { connectionCompressionOptions = PermessageDeflateCompression defaultPermessageDeflate }

reanimate :: Animation -> IO ()
reanimate animation = do
  args <- getArgs
  hSetBuffering stdin NoBuffering
  case args of
    ["once"] -> renderSvgs animation
    _ -> withTempDir $ \tmpDir -> do
      url <- getDataFileName "viewer/build/index.html"
      runCmd "xdg-open" [url]
      runServerWith "127.0.0.1" 9161 opts $ \pending -> do
        putStrLn "Server pending."
        prog <- getProgName
        lst <- listDirectory "."
        mbSelf <- findFile ("." : lst) prog
        blocker <- newEmptyMVar :: IO (MVar ())
        case mbSelf of
          Nothing -> do
            hPutStrLn stderr "Failed to find own source code."
          Just self -> withINotify $ \notify -> do
            conn <- acceptRequest pending
            slave <- newEmptyMVar
            let handler = modifyMVar_ slave $ \tid -> do
                  sendTextData conn (T.pack "Compiling")
                  putStrLn "Kill and respawn."
                  killThread tid
                  tid <- forkIO $ withTempFile ".exe" $ \tmpExecutable -> do
                    ret <- runCmd_ "stack" $ ["ghc", "--"] ++ ghcOptions tmpDir ++ [self, "-o", tmpExecutable]
                    case ret of
                      Left err ->
                        sendTextData conn $ T.pack $ "Error" ++ unlines (drop 3 (lines err))
                      Right{} -> do
                        getFrame <- runCmdLazy tmpExecutable ["once", "+RTS", "-N", "-M200M", "-RTS"]
                        flip fix [] $ \loop acc -> do
                          frame <- getFrame
                          case frame of
                            Left "" -> do
                              sendTextData conn (T.pack "Done")
                              -- insertCache msg (reverse acc)
                            Left err -> do
                              -- _ <- getChanContents queue
                              sendTextData conn $ T.pack $ "Error" ++ err
                            Right frame -> do
                              sendTextData conn frame
                              loop (frame : acc)
                  return tid
            putStrLn "Found self. Listening."
            addWatch notify [Modify] self (const handler)
            putMVar slave =<< forkIO (return ())
            let loop = do
                  fps <- receiveData conn :: IO T.Text
                  handler
                  loop
            loop

ghcOptions :: FilePath -> [String]
ghcOptions tmpDir =
    ["-rtsopts", "--make", "-threaded", "-O2"] ++
    ["-odir", tmpDir, "-hidir", tmpDir]
