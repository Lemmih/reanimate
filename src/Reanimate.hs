{-|
Module      : Reanimate
Description : SVG-base animation library.
Copyright   : Written by David Himmelstrup
License     : Unlicense
Maintainer  : lemmih@gmail.com
Stability   : experimental
Portability : POSIX

Reanimate is an animation library based on SVGs. It is designed to act like glue
between external components such as 'latex', 'ffmpeg', 'gnuplot', 'diagrams',
and 'povray'.

-}
module Reanimate
  ( -- * Driver
    --
    -- | Reanimate features a web-based viewer which is opened by default if
    --   no other parameters are given. Key features:
    --
    --   * This viewer listens for changes to the source file and recompiles the
    --     code automatically as needed.
    --   * Animations are rendered with increasing fidelity until the frame
    --     rate eaches 60 fps.
    --   * Key commands for pausing, frame stepping, forward/rewind.
    reanimate,
    -- * Animations
    Animation(..),
    mkAnimation,
    animate,
    duration,
    -- ** Composition
    before,
    sim,
    simLoop,
    simDrop,
    pause,
    andThen,
    mapA,
    pauseAtEnd,
    pauseAtBeginning,
    pauseAround,
    adjustDuration,
    setDuration,
    reverseA,
    playThenReverseA,
    repeatA,
    freezeAtPercentage,
    -- ** Signals
    Signal,
    signalFlat,
    signalLinear,
    signalFromTo,
    signalReverse,
    signalCurve,
    signalBell,
    signalOscillate,
    signalFromList,
    -- * SVG
    module Reanimate.Svg.Constructors,
    module Reanimate.Svg.LineCommand,
    module Reanimate.Svg.BoundingBox,
    module Reanimate.Svg,
    -- ** Raster data
    embedImage,
    embedDynamicImage,
    embedPng,
    raster,
    -- ** External SVG providers
    latex,
    latexAlign,
    xelatex,
    povray,
    -- * Colormaps
    turbo,
    viridis,
    magma,
    inferno,
    plasma,
    sinebow,
    parula,
    cividis,
    jet,
    hsv,
    hsvMatlab,
    greyscale,
    -- * Constants
    screenWidth,
    screenHeight,
    defaultDPI,
    defaultStrokeWidth
  ) where

import           Reanimate.Animation
import           Reanimate.ColorMap
import           Reanimate.Constants
import           Reanimate.Driver
import           Reanimate.LaTeX
import           Reanimate.Povray
import           Reanimate.Raster
import           Reanimate.Signal
import           Reanimate.Svg
import           Reanimate.Svg.BoundingBox
import           Reanimate.Svg.Constructors
import           Reanimate.Svg.LineCommand
