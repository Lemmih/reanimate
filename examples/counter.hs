#!/usr/bin/env stack
-- stack --resolver lts-13.14 runghc --package reanimate
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards #-}
module Main (main) where

import           Control.Lens
import           Data.Text (pack, Text)
import Numeric

import           Graphics.SvgTree hiding (Text)
import           Reanimate.Driver (reanimate)
import           Reanimate.LaTeX
import           Reanimate.Monad
import           Reanimate.Svg
import           Reanimate.Signal

main :: IO ()
main = reanimate $ mkAnimation 2 $ do
  s <- getSignal $ signalFromTo 0 (2*60-1) signalLinear
  emit $ mkCircle (Num s)
