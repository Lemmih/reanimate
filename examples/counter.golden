0<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle /></g></svg>
1<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="2.428571" /></g></svg>
2<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="4.857142" /></g></svg>
3<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="7.285714" /></g></svg>
4<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="9.714285" /></g></svg>
5<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="12.142857" /></g></svg>
6<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="14.571428" /></g></svg>
7<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="17.0" /></g></svg>
8<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="19.428571" /></g></svg>
9<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="21.857142" /></g></svg>
10<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="24.285714" /></g></svg>
11<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="26.714285" /></g></svg>
12<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="29.142857" /></g></svg>
13<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="31.571428" /></g></svg>
14<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="34.0" /></g></svg>
15<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="36.428571" /></g></svg>
16<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="38.857142" /></g></svg>
17<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="41.285714" /></g></svg>
18<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="43.714285" /></g></svg>
19<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="46.142857" /></g></svg>
20<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="48.571428" /></g></svg>
21<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="51.0" /></g></svg>
22<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="53.428571" /></g></svg>
23<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="55.857142" /></g></svg>
24<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="58.285714" /></g></svg>
25<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="60.714285" /></g></svg>
26<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="63.142857" /></g></svg>
27<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="65.571428" /></g></svg>
28<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="68.0" /></g></svg>
29<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="70.428571" /></g></svg>
30<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="72.857142" /></g></svg>
31<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="75.285714" /></g></svg>
32<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="77.714285" /></g></svg>
33<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="80.142857" /></g></svg>
34<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="82.571428" /></g></svg>
35<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="84.999999" /></g></svg>
36<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="87.428571" /></g></svg>
37<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="89.857142" /></g></svg>
38<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="92.285714" /></g></svg>
39<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="94.714285" /></g></svg>
40<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="97.142857" /></g></svg>
41<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="99.571428" /></g></svg>
42<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="102.0" /></g></svg>
43<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="104.428571" /></g></svg>
44<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="106.857142" /></g></svg>
45<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="109.285714" /></g></svg>
46<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="111.714285" /></g></svg>
47<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="114.142857" /></g></svg>
48<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="116.571428" /></g></svg>
49<svg viewBox="-8.0 -4.5 16.0 9.0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"><g stroke-width="0.05" transform="scale(1.0, -1.0)"><circle r="118.999999" /></g></svg>
